### PR TITLE
ast+topdown: cache array groundness

### LINKED ISSUE #3679: ast.Array should cache groundness bit to avoid unnecessary computation
The ast.Array structure does not memoize groundness. As a result, checks for groundness have to recompute for all elements which can be expensive for large arrays.

Related: 00be8a27cb189f4a732ddbd51546088e2a348493

### PR DESCRIPTION
**Pull request by srenatus**
I had been believing that we couldn't do this, since many methods _allow_ changing an array's elements in ways that would invalidate the cached groundness bit. _However_, if we don't do that, we're OK.

To get the tests to pass, the `walk` builtin implementation had implicitly depended on array plugging returning a copy; that's now made explicit.

Fixes #3679.

### CODE REVIEW

File: topdown/topdown_bench_test.go [hunk 1]
```diff
@@ -29,6 +29,54 @@ func BenchmarkArrayIteration(b *testing.B) {
...
+	for _, n := range sizes {
+		b.Run(fmt.Sprint(n), func(b *testing.B) {
+			data := make([]interface{}, n)
+			for i := 0; i < n; i++ {
+				data[i] = fmt.Sprintf("whatever%d", i)
+			}
+			store := inmem.NewFromObject(map[string]interface{}{"fixture": data})
+			module := `package test
+			fixture := data.fixture
+			main { x := fixture }`
```
Comment (srenatus) [inline topdown/topdown_bench_test.go:46]: I had to make this benchmark oddly specific to see an effect:

```
name                    old time/op    new time/op    delta
ArrayPlugging/10-16       14.4µs ± 1%    14.3µs ± 3%     ~     (p=0.690 n=5+5)
ArrayPlugging/100-16      22.7µs ±11%    21.0µs ±14%     ~     (p=0.095 n=5+5)
ArrayPlugging/1000-16     95.4µs ± 8%    77.8µs ± 5%  -18.41%  (p=0.008 n=5+5)
ArrayPlugging/10000-16     852µs ± 4%     685µs ± 2%  -19.55%  (p=0.008 n=5+5)

name                    old alloc/op   new alloc/op   delta
ArrayPlugging/10-16       9.83kB ± 0%    9.71kB ± 0%   -1.22%  (p=0.008 n=5+5)
ArrayPlugging/100-16      15.1kB ± 0%    14.1kB ± 0%   -6.22%  (p=0.008 n=5+5)
ArrayPlugging/1000-16     65.7kB ± 0%    57.5kB ± 0%  -12.54%  (p=0.008 n=5+5)
ArrayPlugging/10000-16     574kB ± 0%     492kB ± 0%  -14.30%  (p=0.008 n=5+5)

name                    old allocs/op  new allocs/op  delta
ArrayPlugging/10-16          191 ± 0%       188 ± 0%   -1.57%  (p=0.008 n=5+5)
ArrayPlugging/100-16         371 ± 0%       368 ± 0%   -0.81%  (p=0.008 n=5+5)
ArrayPlugging/1000-16      2.17k ± 0%     2.17k ± 0%   -0.14%  (p=0.008 n=5+5)
ArrayPlugging/10000-16     20.2k ± 0%     20.2k ± 0%   -0.01%  (p=0.008 n=5+5)

```

File: topdown/bindings.go [hunk 2]
```diff
@@ -104,6 +107,9 @@ func (u *bindings) plugNamespaced(a *ast.Term, caller *bindings) *ast.Term {
 		})
 		return &cpy
 	case ast.Set:
```
Comment (srenatus) [inline topdown/bindings.go:109]: For the other three shortcuts added here, I have not yet thought about when they'd happen -- they do seem to make sense in a pattern-match kind of way 🙃

**Review by tsandall — APPROVED** [review]
LGTM. The TODOs on the bindings look fine to me.

**Comment by srenatus** [comment]
OK great. I'll update this and get it merged.

**Pull request merged**